app/assets/javascripts/countries/views/pantropical/vis.js

```diff
@@ -372,29 +429,281 @@ function addCommas(nStr) {
+      var diff_years = Math.abs(this.year_right - this.year_left);
+      var years_between = diff_years - 1;
+      var years_total = diff_years + 2;
+      var lookup_years = [];
+
+      // push years onto array
+
+      lookup_years.push(year_left);
+      // generate years in between
+      for(var i = 0; i < years_between; i++){
+        var sum = i + 1;
+        lookup_years.push(year_left + sum);
+      }
+      lookup_years.push(year_right);
+
+      // look up data corresponding to year
+      var sum_data = 0;
+      for (year in lookup_years){
+        for (key in d) {
+          if (key == ("y" + lookup_years[year].toString())){
+            sum_data += parseFloat(d[key]);
+          }
+        }
+      }
+
+      var avg = sum_data/years_total;
+
+      return avg.toFixed(6);
+    };
+
+    // returns array of bubbles in descending order
+    // array [{id1, value1}, {id2, value2}, ...]
+    BubbleChart.prototype.sort_desc = function(array){
+
+      array.sort(function(a, b) {
+        if (a.value > b.value) {
+          return 1;
+        }
+        if (a.value < b.value) {
+          return -1;
+        }
+        return 0;
+      });
+      array = array.reverse();
+
+      return array;
+    };
+
+    BubbleChart.prototype.set_years = function(year){
+      this.year = year;
+
+      if (! !!this.year) {
+        this.year_left = 2001;
+        this.year_right = 2001;
+      } else {
+        this.year_left = this.year[0];
+        this.year_right = this.year[1];
+      }
+    };
+
+
+    BubbleChart.prototype.get_values_array = function() {
+      var that = this;
+      var values_array = [];
+
+      this.circles.each(function(d) {
+        if (that.year_left == that.year_right) {
+          var year_string = "y" + that.year_left;
+          var value = d[year_string];
+        } else {
+          var value = that.calculate_average(d);
+        }
+        var id = ~~d.id;
+        values_array.push({id: id, value: value});
+      });
+      return this.sort_desc(values_array);
+    };
+
+    // returns an array of the coordinates according to the sorted_array index (bubble_id, starts at 1)
+    // sorted_index, index_of start at 0
+    BubbleChart.prototype.get_coordinates = function(sorted_index){
+      var dist_x = 125;
+      var dist_y = 55;
+      var offset_x = 150;
+      var offset_y = 50;
+      var col_count = 6;
+
+      if (sorted_index < 4){
+        // bubbles at index 0, 1, 2, 3 are in line 1
+        col_count = 4;
+        dist_x = 170;
+        offset_x = 275;
+      } else {
+        // modify current index to push bubbles into line 2
+        sorted_index += 2;
+      }
+
+      if (sorted_index === 0 ) {
+        offset_x = 225
+      }
+      
+      if (sorted_index === 3 ) {
+        offset_x = 260
+      }
+
+      var x_position = (((sorted_index) % col_count) * dist_x);
+      var y_position = 0;
+
+      if (sorted_index < (col_count)){
+        y_position = 150;
```

Comment:
TO-DO: in the future, we need to get rid of this nonsense, module operator or a loop or a function or whatever :) /cc @paulahaertel